x-pack/plugins/endpoint/public/applications/endpoint/store/policy_list/test_mock_utils.ts

```diff
@@ -0,0 +1,120 @@
+  mockedHttpService: jest.Mocked<HttpStart>,
+  responseItems: GetDatasourcesResponse['items'] = [generator.generatePolicyDatasource()]
+): void => {
+  mockedHttpService.get.mockImplementation((...args) => {
+    const [path] = args;
+    if (typeof path === 'string') {
+      if (path === INGEST_API_DATASOURCES) {
+        return Promise.resolve<GetDatasourcesResponse>({
+          items: responseItems,
+          total: 10,
+          page: 1,
+          perPage: 10,
+          success: true,
+        });
+      }
+    }
+    return Promise.reject(new Error(`MOCK: unknown policy list api: ${path}`));
+  });
+};
+
+/**
+ * Utilities for testing Redux middleware
+ */
+export interface MiddlewareActionSpyHelper<S = GlobalState> {
+  /**
+   * Returns a promise that is fulfilled when the given action is dispatched or a timeout occurs.
+   * The use of this method instead of a `sleep()` type of delay should avoid test case instability
+   * especially when run in a CI environment.
+   *
+   * @param actionType
+   */
+  waitForAction: (actionType: Pick<AppAction, 'type'>['type']) => Promise<void>;
+  /**
+   * Redux middleware that enables spying on the action that are dispatched through the store
+   */
+  actionSpyMiddleware: ReturnType<MiddlewareFactory<S>>;
+}
+
+/**
+ * Creates a new instance of middleware action helpers
+ * Note: in most cases (testing concern specific middleware) this function should be given
+ * the state type definition, else, the global state will be used.
+ *
+ * @example
+ * // Use in Policy List middleware testing
+ * const { actionSpyMiddleware, waitForAction } = createSpyMiddleware<PolicyListState>();
+ * store = createStore(
+ *    policyListReducer,
+ *    applyMiddleware(
+ *      policyListMiddlewareFactory(fakeCoreStart, depsStart),
+ *      actionSpyMiddleware
+ *    )
+ *  );
+ * // later in test
+ * it('...', async () => {
+ *   //...
+ *   await waitForAction('serverReturnedPolicyListData');
+ *   // do assertion
+ * });
+ */
+export const createSpyMiddleware = <S = GlobalState>(): MiddlewareActionSpyHelper<S> => {
+  type ActionWatcher = (action: AppAction) => void;
+
+  const watchers = new Set<ActionWatcher>();
+
+  return {
+    waitForAction: async (actionType: string) => {
+      // Error is defined here so that we get a better stack trace that points to the test from where it was used
+      const err = new Error(`action '${actionType}' was not dispatched within the allocated time`);
+
+      await new Promise((resolve, reject) => {
+        const watch: ActionWatcher = action => {
+          if (action.type === actionType) {
+            watchers.delete(watch);
+            clearTimeout(timeout);
+            resolve();
+          }
+        };
+
+        // We timeout before jest's default 5s, so that a better error stack is returned
+        const timeout = setTimeout(() => {
+          watchers.delete(watch);
+          reject(err);
+        }, 4500);
```

Comment:
would it ever make sense to make this timeout configurable?  Not necessary now, but wondering if we would ever adjust jest's default.

Reply:
Yeah, for sure the `createSpyMiddleware` factory can be improved in the future to take in the timeout. I actually looked to see if there was a way to access Jest's timeout value at runtime so that I could use it, but was not able to find it (there is one to set it, but not retrieve it).

Reply:
@paul-tavares what happens if you dont have your own timeout? wouldn't jest timeout at its configured time anyway?